crates/primitives/src/receipt.rs

```diff
@@ -70,6 +69,105 @@ impl Receipt {
     pub fn with_bloom_ref(&self) -> ReceiptWithBloomRef<'_> {
         self.into()
     }
+
+    /// Returns length of RLP-encoded receipt fields with the given [`Bloom`] without an RLP header.
+    pub fn rlp_encoded_fields_length_with_bloom(&self, bloom: &Bloom) -> usize {
+        self.cumulative_gas_used.length() + bloom.length() + self.logs.length()
+    }
+
+    /// RLP-encodes receipt fields with the given [`Bloom`] without an RLP header.
+    pub fn rlp_encode_fields_with_bloom(&self, bloom: &Bloom, out: &mut dyn BufMut) {
+        self.success.encode(out);
+        self.cumulative_gas_used.encode(out);
+        bloom.encode(out);
+        self.logs.encode(out);
+    }
+
+    /// Returns RLP header for this receipt encoding with the given [`Bloom`].
+    pub fn rlp_header_with_bloom(&self, bloom: &Bloom) -> Header {
+        Header { list: true, payload_length: self.rlp_encoded_fields_length_with_bloom(bloom) }
+    }
+
+    /// RLP-decodes receipt's field with a [`Bloom`].
+    ///
+    /// Does not expect an RLP header.
+    pub fn rlp_decode_fields_with_bloom(
+        buf: &mut &[u8],
+    ) -> alloy_rlp::Result<ReceiptWithBloom<Self>> {
+        let tx_type = Decodable::decode(buf)?;
+        let success = Decodable::decode(buf)?;
+        let cumulative_gas_used = Decodable::decode(buf)?;
+        let logs_bloom = Decodable::decode(buf)?;
+        let logs = Decodable::decode(buf)?;
+
+        Ok(ReceiptWithBloom {
+            receipt: Self {
+                tx_type,
+                success,
+                cumulative_gas_used,
+                logs,
+                #[cfg(feature = "optimism")]
+                deposit_nonce: if tx_type == TxType::Deposit {
+                    Some(Decodable::decode(buf)?)
+                } else {
+                    None
+                },
+                #[cfg(feature = "optimism")]
+                deposit_receipt_version: if tx_type == TxType::Deposit {
+                    Some(Decodable::decode(buf)?)
+                } else {
+                    None
+                },
+            },
+            logs_bloom,
+        })
+    }
+}
+
+impl RlpEncodableReceipt for Receipt {
+    fn rlp_encoded_length_with_bloom(&self, bloom: &Bloom) -> usize {
+        self.rlp_header_with_bloom(bloom).length_with_payload()
+    }
+
+    fn rlp_encode_with_bloom(&self, bloom: &Bloom, out: &mut dyn BufMut) {
+        self.rlp_header_with_bloom(bloom).encode(out);
+        self.rlp_encode_fields_with_bloom(bloom, out);
+    }
```

Comment:
this is wrong, the rlp encoding should be equivalent to `ReceiptWithBloomEncoder::encode_inner(... true)` and EIP-2718 should be equivalent to `encode_inner(..., false)`

Reply:
i have now removed the `ReceiptWithBloomEncoder` struct and transferred some functions to the receipt impl. But I believe I could instead shorten the `Receipt` impl block as well and directly use `ReceiptWithBloom` encoding functions? But then this `Receipt` as extra fields for OP and also `tx_type`.. What do you say?

Reply:
Also, shall I try removing `ReceiptWithBloomRef` struct as well? Looking to shorten the code.

Reply:
> i have now removed the `ReceiptWithBloomEncoder` struct and transferred some functions to the receipt impl. But I believe I could instead shorten the `Receipt` impl block as well and directly use `ReceiptWithBloom` encoding functions? But then this `Receipt` as extra fields for OP and also `tx_type`.. What do you say?

no, we need to keep the current `Receipt` type as is, and define all encoding on it

`ReceiptWithBloomRef` should be removed now

Reply:
Thanks! Will do.